### PR TITLE
BitFieldArray getters now support some types, add basic/nullable/optional phantom types

### PR DESCRIPTION
**Pull request by npaun**
## Summary
This PR contains a big set of features designed to reduce the amount of boilerplate we write.

* `BitFieldArray` fields now support some types. This is only defined for cases where a static_cast from uint32_t to that type is legal.  The type involved will now automatically be added to the generated header file. This includes:
     - Python native enums (which will cause a header file change without any further action)
     - `namedstruct` phantom types
* For cases where we want to export an extra Python type to C++ even if there are no associated values, I generalized `Struct.addEnumType` to `Struct.addTypeDeclaration`.
* `namedstruct` is now responsible for the `phantom_type.h` header. 
    - There's still a `TransitUtils/phantom_type.h` but it now only includes convenience macros, and specializations for `hash` and `msgpack` [separate PR for this].
    - Unfortunately, I found it impossible to do a forward declaration for msgpack stuff so there's now an compile option `-DNAMEDSTRUCT_DECLARE_MSGPACK` that will hackily add those prototypes.
* Added `phantom_optional`. This is like `nullable_phantom_type` but it does not act as a numerical type. You must explicitly use `has_value` and `operator*`, like a `std::optional`.
   - Using phantom optional requires a signed type as nullopt is stored as `-1`. Open to suggestions for doing this differently.
* Added `OptionalGet` which does null coalescing and type inference for Python:
    - e.g. `OptionalGet(some_entity).parent_entity.parent_index`
         - Will store 0 if either `parent_entity` or `parent_index` are null, `1 + parent_index` if set.
         - Will derive the type `phantom_optional<ParentId>` even if we never encounter a non-null value for `parent_entity` while generating a header.
         - See below for the nitty gritty configuration to achieve this.

## How to enable `OptionalGet`.

* Type annotate the entities

```python
from typing import Optional # Python 3 type hints

class CompressorSomeEntity(Entity):
     parent: Optional[ParentEntity]

class ParentEntity(Entity):
     parent_index: ParentId
```

* Declare a Python "phantom type" (distinct name, but no safety unlike C++):

```python

class ParentId(int):
   pass
```

(Note for @hausen: I was mistaken earlier when I said we needed to re-implement `__new__` - for a simple alias like this it'll automatically find `int`'s method.)

## Next steps

* It's probably quite easy to generate the `getXId(XStruct const * const xStruct, XId xId)` non-member functions as well. We probably don't need to write boilerplate for most of these. But that's for another time.

### CODE REVIEW

File: include/namedstruct/phantom_type.h [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+#define namedstruct_phantom_type_h
+
+#include <limits>
+#include <type_traits>
+#include <utility>
+#include <optional>
+
+#ifdef NAMEDSTRUCT_DECLARE_MSGPACK
+#include <TransitUtils/msgpack_support.h>
+#endif
```
Comment (hausen) [inline include/namedstruct/phantom_type.h:35]: `phantom_type.h` should have an optional dependency on `msgpack` only:
```suggestion
#ifdef NAMEDSTRUCT_MSGPACK_SUPPORT
#include <msgpack.hpp>
#endif
```

File: include/namedstruct/phantom_type.h [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    template <typename Other>
+    inline constexpr auto as() const {
+        return static_cast<Other>(value);
+    }
+
+#ifdef NAMEDSTRUCT_DECLARE_MSGPACK
+    MSGPACK_SUPPORT
+#endif
```
Comment (hausen) [inline include/namedstruct/phantom_type.h:140]:
```suggestion
#ifdef NAMEDSTRUCT_MSGPACK_SUPPORT
    void msgpack_pack(msgpack::packer<msgpack::sbuffer>& pk) const {
        ...
    }

    void msgpack_unpack(const msgpack::object& o) {
        ....
    }
#endif
```

File: include/namedstruct/phantom_type.h [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        return static_cast<Other>(phantom_type<T,Derived>::value);
+    }
+};
+
+/** A phantom type which enforces the use of has_value or dereference. Automatically
+ * decoded from the namedstruct representation
+ *  nullopt: 0
+ *  value: 1 + value
+ */
+template <typename PhantomType> struct phantom_optional {
```
Comment (hausen) [inline include/namedstruct/phantom_type.h:175]: is this better than simply using an `std::optional<PhantomType>` ?

File: include/namedstruct/phantom_type.h [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    static constexpr PhantomType NullValue = PhantomType(-1);
+    static constexpr PhantomType ValueShift = PhantomType(1);
```
Comment (hausen) [inline include/namedstruct/phantom_type.h:177]: I'd make these template args with default values.

File: include/namedstruct/phantom_type.h (hunk 3 above)
Comment (npaun) [reply include/namedstruct/phantom_type.h:175]: If we want to simplify the code,  we can just implement this class by inheriting from std::optional and doing the "-1 or nullopt" in the constructor. 

I suspect but have no evidence that using a sentinel value for nullopt is a bit better storage and perf wise vs a tagged union (which I'm guessing is how std optional is built()

**Comment by npaun** [comment]
Ready for re-review

File: include/namedstruct/phantom_type.h [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+    template <typename Other>
+    inline constexpr auto as() const {
+        return static_cast<Other>(value);
+    }
+
+#ifdef NAMEDSTRUCT_MSGPACK_SUPPORT
+    void msgpack_pack(msgpack::packer<msgpack::sbuffer>& pk) const;
+
+    void msgpack_unpack(const msgpack::object& o);
```
Comment (hausen) [inline include/namedstruct/phantom_type.h:141]: Sorry, I wrongly said it was ok to merely declare the method stubs, but this won't work if the method bodies aren't defined:
```suggestion
    void msgpack_pack(msgpack::packer<msgpack::sbuffer>& pk) const {
        pk.pack(value);
    }

    void msgpack_unpack(const msgpack::object& o) {
        value = o.as<value_type>();
    }
```

**Review by hausen — APPROVED** [review]

**Pull request merged**